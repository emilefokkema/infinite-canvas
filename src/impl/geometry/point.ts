export class Point{
	constructor(public readonly x: number, public readonly y: number){}
	public mod(): number{
		return Math.sqrt(this.modSq());
	}
	public modSq(): number{
		return this.x * this.x + this.y * this.y;
	}
	public minus(other: Point): Point{
		return new Point(this.x - other.x, this.y - other.y);
	}
	public plus(other: Point): Point{
		return new Point(this.x + other.x, this.y + other.y);
	}
	public dot(other: Point): number{
		return this.x * other.x + this.y * other.y;
	}
	public cross(other: Point): number{
		return this.x * other.y - this.y * other.x;
	}
	public equals(other: Point): boolean{
		return this.x === other.x && this.y === other.y;
	}
	public getPerpendicular(): Point{
		return new Point(-this.y, this.x);
	}
	public scale(r: number): Point{
		return new Point(r * this.x, r * this.y);
	}
	public projectOn(other: Point): Point{
		return other.scale(this.dot(other) / other.modSq());
	}
	public matrix(a: number, b: number, c: number, d: number): Point{
		return new Point(a * this.x + b * this.y, c * this.x + d * this.y);
	}
	public inSameDirectionAs(other: Point): boolean{
		return this.cross(other) === 0 && this.dot(other) >= 0;
	}
	public isInOppositeDirectionAs(other: Point): boolean{
		return this.cross(other) === 0 && this.dot(other) < 0;
	}
	public isOnSameSideOfOriginAs(other1: Point, other2: Point){
		return this.isInSmallerAngleBetweenPoints(other1, other2) || 
			   other1.isInSmallerAngleBetweenPoints(this, other2) ||
			   other2.isInSmallerAngleBetweenPoints(this, other1);
	}
	public isInSmallerAngleBetweenPoints(point1: Point, point2: Point): boolean{
		const cross: number = point1.cross(point2);
		if(cross > 0){
			return point1.cross(this) >= 0 && this.cross(point2) >= 0;
		}else if(cross < 0){
			return point1.cross(this) <= 0 && this.cross(point2) <= 0;
		}else if(point1.dot(point2) > 0){
			return this.cross(point1) === 0 && this.dot(point1) > 0;
		}else{
			return true;
		}
	}
	public static origin: Point = new Point(0, 0);
}