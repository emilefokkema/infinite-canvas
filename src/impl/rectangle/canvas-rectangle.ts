import { ConvexPolygon } from "../areas/polygons/convex-polygon";
import { Point } from "../geometry/point";
import { CoordinateSystem } from "./coordinate-system";
import { CanvasMeasurement } from "./canvas-measurement";
import { Transformation } from "../transformation";
import { TransformationRepresentation } from "api/transformation-representation";
import { Units } from "api/units";

export interface CanvasRectangle{
    readonly viewboxWidth: number;
    readonly viewboxHeight: number;
    readonly polygon: ConvexPolygon;
    readonly userTransformation: Transformation;
    readonly infiniteCanvasContext: CoordinateSystem;
    readonly initialBitmapTransformation: Transformation;
    addPathAroundViewbox(context: CanvasRenderingContext2D, margin: number, counterclockwise: boolean): void;
    getCSSPosition(clientX: number, clientY: number): Point;
    getTransformationForInstruction(infiniteCanvasContextTransformation: TransformationRepresentation): Transformation;
    translateInfiniteCanvasContextTransformationToBitmapTransformation(infiniteCanvasContextTransformation: TransformationRepresentation): Transformation;
    getBitmapTransformationToTransformedInfiniteCanvasContext(): Transformation;
    getBitmapTransformationToInfiniteCanvasContext(): Transformation;
    withMeasurement(measurement: CanvasMeasurement): CanvasRectangle
    withTransformation(transformation: TransformationRepresentation): CanvasRectangle
    withUnits(units: Units): CanvasRectangle
}