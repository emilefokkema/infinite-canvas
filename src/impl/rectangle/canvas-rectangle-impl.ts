import { CanvasRectangle } from './canvas-rectangle';
import { CoordinatesSwitch } from './coordinates-switch';
import { CanvasMeasurement } from "./canvas-measurement";
import { ConvexPolygon } from "../areas/polygons/convex-polygon";
import { Point } from "../geometry/point";
import { TransformationRepresentation } from "api/transformation-representation";
import { Transformation } from "../transformation";
import { Units } from "api/units";
import { CoordinateSystem } from './coordinate-system';
import { getRectWithFourEdges } from '../rect/get-rect-strategy';

function measurementsAreOfEqualSize(one: CanvasMeasurement, other: CanvasMeasurement): boolean{
    if(!one){
        return !other;
    }
    if(!other){
        return false;
    }
    return one.viewboxWidth === other.viewboxWidth &&
            one.viewboxHeight === other.viewboxHeight &&
            one.screenWidth === other.screenWidth &&
            one.screenHeight === other.screenHeight;
}

export class CanvasRectangleImpl implements CanvasRectangle{
    public get viewboxWidth(): number{return this.measurement.viewboxWidth;}
    public get viewboxHeight(): number{return this.measurement.viewboxHeight}
    public get userTransformation(): Transformation{return this.coordinates.userTransformation;}
    public get infiniteCanvasContext(): CoordinateSystem{return this.coordinates.infiniteCanvasContext;}
    public get initialBitmapTransformation(): Transformation{return this.coordinates.initialBitmapTransformation;}
    constructor(
        private readonly coordinates: CoordinatesSwitch,
        private readonly measurement: CanvasMeasurement,
        public readonly polygon: ConvexPolygon){
    }

    public withUnits(units: Units): CanvasRectangle{
        const coordinatesSwitch = this.coordinates.withUnits(units);
        return new CanvasRectangleImpl(coordinatesSwitch, this.measurement, this.polygon)
    }

    public withTransformation(transformation: TransformationRepresentation): CanvasRectangle{
        const coordinatesSwitch = this.coordinates.withUserTransformation(Transformation.create(transformation));
        return new CanvasRectangleImpl(coordinatesSwitch, this.measurement, this.polygon);
    }

    public withMeasurement(measurement: CanvasMeasurement): CanvasRectangle{
        const isEqual = measurementsAreOfEqualSize(this.measurement, measurement);
        if(isEqual){
            return this;
        }
        const {viewboxWidth, viewboxHeight} = measurement;
        const polygon = getRectWithFourEdges(0, 0, viewboxWidth, viewboxHeight).getArea();
        const coordinatesSwitch = this.coordinates.withCanvasMeasurement(measurement);
        return new CanvasRectangleImpl(coordinatesSwitch, measurement, polygon)
    }

    public getCSSPosition(clientX: number, clientY: number): Point{
        const {left, top} = this.measurement
        return new Point(clientX - left, clientY - top);
    }

    public getTransformationForInstruction(infiniteCanvasContextTransformation: TransformationRepresentation): Transformation{
        return this.coordinates.getTransformationForInstruction(infiniteCanvasContextTransformation)
    }

    public translateInfiniteCanvasContextTransformationToBitmapTransformation(infiniteCanvasContextTransformation: TransformationRepresentation): Transformation{
        return this.coordinates.translateInfiniteCanvasContextTransformationToBitmapTransformation(infiniteCanvasContextTransformation)
    }

    public getBitmapTransformationToTransformedInfiniteCanvasContext(): Transformation{
        return this.coordinates.getBitmapTransformationToTransformedInfiniteCanvasContext();
    }

    public getBitmapTransformationToInfiniteCanvasContext(): Transformation{
        return this.coordinates.getBitmapTransformationToInfiniteCanvasContext();
    }

    public addPathAroundViewbox(context: CanvasRenderingContext2D, margin: number): void{
        const width: number = this.viewboxWidth + 2 * margin;
        const height: number = this.viewboxHeight + 2 * margin;
        context.save();
        context.setTransform(1, 0, 0, 1, 0, 0);
        context.rect(-margin, -margin, width, height);
        context.restore();
    }

    public static create(measurement: CanvasMeasurement, units: Units): CanvasRectangleImpl{
        const {viewboxWidth, viewboxHeight} = measurement;
        const polygon = getRectWithFourEdges(0, 0, viewboxWidth, viewboxHeight).getArea();
        const coordinates = CoordinatesSwitch.create(units, measurement);
        return new CanvasRectangleImpl(coordinates, measurement, polygon)
    }
}