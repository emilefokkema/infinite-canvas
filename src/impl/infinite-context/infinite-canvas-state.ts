import { ViewBox } from "../interfaces/viewbox";

export class InfiniteCanvasState implements CanvasState{
	constructor(private viewBox: ViewBox){}
	public restore():void{
		this.viewBox.restoreState();
	}
	public save(): void{
		this.viewBox.saveState();
	}
	public reset(): void{
		this.viewBox.resetState();
	}
}