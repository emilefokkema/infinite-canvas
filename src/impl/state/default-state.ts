import { InfiniteCanvasState } from "./infinite-canvas-state";
import { InfiniteCanvasStateInstance } from "./infinite-canvas-state-instance";

export const defaultState: InfiniteCanvasState = new InfiniteCanvasState(InfiniteCanvasStateInstance.default, []);