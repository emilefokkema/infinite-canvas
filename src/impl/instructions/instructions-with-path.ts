import { StateChangingInstructionSequence } from "./state-changing-instruction-sequence";
import { PreExecutableInstructionWithState } from "./pre-executable-instruction-with-state";
import { CurrentPath } from "../interfaces/current-path";
import { PathInstruction } from "../interfaces/path-instruction";
import { InfiniteCanvasState } from "../state/infinite-canvas-state";
import { Instruction } from "./instruction";
import { Area } from "../areas/area";
import { InfiniteCanvasAreaBuilder } from "../areas/infinite-canvas-area-builder";
import { Position } from "../geometry/position";
import { transformPosition } from "../geometry/transform-position";
import { InstructionsWithSubpath } from "./instructions-with-subpath";
import { DrawnPathProperties } from "../interfaces/drawn-path-properties";
import { InfiniteCanvasPathInfinityProvider } from "../infinite-canvas-path-infinity-provider";
import { InstructionsToClip } from "../interfaces/instructions-to-clip";
import { InstructionsToClipImpl } from "./instructions-to-clip-impl";
import { ExecutableStateChangingInstructionSet } from "../interfaces/executable-state-changing-instruction-set";
import { ExecutableStateChangingInstructionSequence } from "./executable-state-changing-instruction-sequence";
import { ExecutableInstructionWithState } from "./executable-instruction-with-state";

export class InstructionsWithPath extends StateChangingInstructionSequence<InstructionsWithSubpath> implements CurrentPath{
    private areaBuilder: InfiniteCanvasAreaBuilder = new InfiniteCanvasAreaBuilder();
    constructor(private _initiallyWithState: PreExecutableInstructionWithState){
        super(_initiallyWithState);
    }
    public get area(): Area{return this.areaBuilder.area;}
    public surroundsFinitePoint(): boolean{
        for(const subpath of this.added){
            if(subpath.surroundsFinitePoint()){
                return true;
            }
        }
        return false;
    }
    public currentSubpathIsClosable(): boolean{
        if(this.added.length === 0){
            return true;
        }
        return this.added[this.added.length - 1].isClosable();
    }
    public allSubpathsAreClosable(): boolean{
        if(this.added.length === 0){
            return true;
        }
        for(const subPath of this.added){
            if(!subPath.isClosable()){
                return false;
            }
        }
        return true;
    }
    public drawPath(instruction: Instruction, state: InfiniteCanvasState, drawnPathProperties: DrawnPathProperties): ExecutableStateChangingInstructionSet{
        if(this.added.length === 0){
            return;
        }
        const toAdd = ExecutableInstructionWithState.create(state, instruction);
        const result = this.makeExecutable(drawnPathProperties);
        toAdd.setInitialState(result.state)
        result.add(toAdd)
        return result;
    }
    private makeExecutable(drawnPathProperties: DrawnPathProperties): ExecutableStateChangingInstructionSequence<ExecutableStateChangingInstructionSet>{
        const result = new ExecutableStateChangingInstructionSequence<ExecutableStateChangingInstructionSet>(this._initiallyWithState.makeExecutable());
        const pathInfinityProvider = new InfiniteCanvasPathInfinityProvider(drawnPathProperties)
        for(const added of this.added){
            result.add(added.makeExecutable(pathInfinityProvider))
        }
        return result;
    }
    private getInstructionsToClip(): InstructionsToClip{
        const executable = this.makeExecutable({lineWidth: 0, lineDashPeriod: 0, shadowOffsets: []})
        executable.setInitialState(executable.stateOfFirstInstruction)
        return new InstructionsToClipImpl(executable, this.area)
    }
    public clipPath(instruction: Instruction, state: InfiniteCanvasState): void{
        if(this.added.length === 0){
            return;
        }
        const currentSubpath: InstructionsWithSubpath = this.added[this.added.length - 1];
        const toAdd: PreExecutableInstructionWithState = PreExecutableInstructionWithState.create(state, instruction);
        currentSubpath.addInstruction(toAdd);
        const instructionsToClip = this.getInstructionsToClip();
        this.addClippedPath(instructionsToClip);
    }
    public closePath(): void{
        if(this.added.length === 0){
            return;
        }
        const currentSubpath: InstructionsWithSubpath = this.added[this.added.length - 1];
        currentSubpath.closePath();
    }
    public moveTo(position: Position, state: InfiniteCanvasState): void{
        const transformedPointToMoveTo: Position = transformPosition(position, state.current.transformation);
        this.areaBuilder.addPosition(transformedPointToMoveTo);
        const newSubpath: InstructionsWithSubpath = InstructionsWithSubpath.create(state, position);
        newSubpath.setInitialState(this.state);
        this.add(newSubpath);
    }
    public canAddLineTo(position: Position, state: InfiniteCanvasState): boolean{
        if(this.added.length === 0){
          return true;
        }
        const transformedPosition: Position = transformPosition(position, state.current.transformation);
        return this.added[this.added.length - 1].canAddLineTo(transformedPosition);
    }
    public lineTo(position: Position, state: InfiniteCanvasState): void{
        if(this.added.length === 0){
            this.moveTo(position, state);
            return;
        }
        const currentSubpath: InstructionsWithSubpath = this.added[this.added.length - 1];
        const transformedPosition: Position = transformPosition(position, state.current.transformation);
        this.areaBuilder.addPosition(transformedPosition);
        currentSubpath.lineTo(position, state);
    }
    public addPathInstruction(pathInstruction: PathInstruction, state: InfiniteCanvasState): void{
        if(this.added.length === 0){
            if(pathInstruction.initialPoint){
                this.moveTo(pathInstruction.initialPoint, state);
            }else{
                return;
            }
        }
        const currentSubpath: InstructionsWithSubpath = this.added[this.added.length - 1];
        const currentPosition: Position = currentSubpath.currentPosition;
        const inverseTransformedCurrentPosition = transformPosition(currentPosition, state.current.transformation.inverse());
        pathInstruction.changeArea(this.areaBuilder.transformedWith(state.current.transformation), inverseTransformedCurrentPosition);
        const toAdd: PreExecutableInstructionWithState = PreExecutableInstructionWithState.create(state, pathInstruction.instruction);
        currentSubpath.addPathInstruction(pathInstruction, toAdd, state);
    }
    public static create(initialState: InfiniteCanvasState): InstructionsWithPath{
        return new InstructionsWithPath(PreExecutableInstructionWithState.create(initialState, (context: CanvasRenderingContext2D) => {context.beginPath();}));
    }
}
