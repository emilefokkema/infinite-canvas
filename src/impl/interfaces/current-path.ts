import { Instruction } from "../instructions/instruction";
import { PathInstruction } from "./path-instruction";
import { StateChangingInstructionSet } from './state-changing-instruction-set'
import { InfiniteCanvasState } from "../state/infinite-canvas-state";
import { Position } from "../geometry/position"
import { DrawablePath } from "./drawable-path";

export interface CurrentPath extends StateChangingInstructionSet, DrawablePath{
    allSubpathsAreClosable(): boolean;
    currentSubpathIsClosable(): boolean;
    surroundsFinitePoint(): boolean;
    clipPath(instruction: Instruction, state: InfiniteCanvasState): void;
    addPathInstruction(pathInstruction: PathInstruction, state: InfiniteCanvasState): void;
    closePath(): void;
    moveTo(position: Position, state: InfiniteCanvasState): void;
    canAddLineTo(position: Position, state: InfiniteCanvasState): boolean;
    lineTo(position: Position, state: InfiniteCanvasState): void;
}