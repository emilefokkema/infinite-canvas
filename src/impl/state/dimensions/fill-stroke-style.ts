import { InfiniteCanvasFillStrokeStyle } from "../../styles/infinite-canvas-fill-stroke-style";
import { Instruction } from "../../instructions/instruction";
import { TypedStateInstanceDimension } from "./typed-state-instance-dimension";
import { InfiniteCanvasStateInstance } from "../infinite-canvas-state-instance";
import { InfiniteCanvasPattern } from "../../styles/infinite-canvas-pattern";

class FillStrokeStyle<K extends "fillStyle" | "strokeStyle"> implements TypedStateInstanceDimension<string | CanvasGradient | CanvasPattern>{
    constructor(private readonly propName: K){}
    public changeInstanceValue(instance: InfiniteCanvasStateInstance, newValue: string | CanvasGradient | CanvasPattern): InfiniteCanvasStateInstance{
        return instance.changeProperty(this.propName, newValue);
    }
    public isEqualForInstances(one: InfiniteCanvasStateInstance, other: InfiniteCanvasStateInstance): boolean{
        return one[this.propName] === other[this.propName];
    }
    public getInstructionToChange(fromInstance: InfiniteCanvasStateInstance, toInstance: InfiniteCanvasStateInstance): Instruction{
        const newValue: string | CanvasGradient | CanvasPattern = toInstance[this.propName];
        if(this.isEqualForInstances(fromInstance, toInstance)){
            if(!(newValue instanceof InfiniteCanvasFillStrokeStyle) || fromInstance.fillAndStrokeStylesTransformed === toInstance.fillAndStrokeStylesTransformed){
                return () => {};
            }
            return toInstance.fillAndStrokeStylesTransformed ? newValue.getInstructionToSetTransformed(this.propName) : newValue.getInstructionToSetUntransformed(this.propName);
        }
        if(newValue instanceof InfiniteCanvasFillStrokeStyle){
            return toInstance.fillAndStrokeStylesTransformed ? newValue.getInstructionToSetTransformed(this.propName) : newValue.getInstructionToSetUntransformed(this.propName);
        }
        

        return (context: CanvasRenderingContext2D) => {context[this.propName] = toInstance[this.propName];};
    }
    public valueIsTransformableForInstance(instance: InfiniteCanvasStateInstance): boolean{
        return !(instance[this.propName] instanceof InfiniteCanvasPattern);
    }
}
export const fillStyle = new FillStrokeStyle("fillStyle");
export const strokeStyle = new FillStrokeStyle("strokeStyle");
