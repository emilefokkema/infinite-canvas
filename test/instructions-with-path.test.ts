import {describe, it, expect, beforeEach } from '@jest/globals';
import { InfiniteCanvasState } from "../src/state/infinite-canvas-state";
import { logWithState } from "./log-with-state";
import { defaultState } from "../src/state/default-state";
import { fillStyle } from "../src/state/dimensions/fill-stroke-style";
import { InstructionsWithPath } from "../src/instructions/instructions-with-path";
import { CurrentPath } from '../src/interfaces/current-path';
import { Point } from "../src/geometry/point";
import { CanvasRectangle } from "../src/rectangle/canvas-rectangle";
import { CanvasRectangleImpl } from "../src/rectangle/canvas-rectangle-impl";
import { MockCanvasMeasurementProvider } from "./mock-canvas-measurement-provider";

function drawAndLog(instructionsWithPath: CurrentPath, state: InfiniteCanvasState): string[]{
    const result = instructionsWithPath.drawPath(
        (context: CanvasRenderingContext2D) => {context.fill();},
        state,
        {
            lineWidth: 0,
            lineDashPeriod: 0,
            shadowOffsets: []
        });
    return logWithState(result);
}

describe("a set of instructions that is also about a path", () => {
    let instructionsWithPath: InstructionsWithPath;
    let currentState: InfiniteCanvasState;
    let rectangle: CanvasRectangle;

    beforeEach(() => {
        rectangle = new CanvasRectangleImpl(new MockCanvasMeasurementProvider(200, 200), {})
        currentState = defaultState;
        instructionsWithPath = InstructionsWithPath.create(currentState, rectangle);
    });

    describe("that receives a change of state", () => {

        beforeEach(() => {
            currentState = currentState.withCurrentState(fillStyle.changeInstanceValue(currentState.current, "#f00"));
        });

        describe("and then receives an instruction that modifies the path", () => {

            beforeEach(() => {
                instructionsWithPath.moveTo(new Point(0, 0), currentState);
            });

            it("should contain an instruction to begin a path, change the state and modify the path", () => {
                expect(drawAndLog(instructionsWithPath, currentState)).toMatchSnapshot();
            });

            describe("and then receives a change of state on the same property as the previous change of state", () => {

                beforeEach(() => {
                    currentState = currentState.withCurrentState(fillStyle.changeInstanceValue(currentState.current, "#00f"));
                });
    
                it("should not have replaced the previous change of state", () => {
                    expect(drawAndLog(instructionsWithPath, currentState)).toMatchSnapshot();
                });

                describe("and then draws the path and receives another change of state on the same property", () => {

                    beforeEach(() => {
                        instructionsWithPath.drawPath((context: CanvasRenderingContext2D) => {
                            context.fill();
                        }, currentState, {lineWidth: 0, lineDashPeriod: 0, shadowOffsets: []});
                        currentState = currentState.withCurrentState(fillStyle.changeInstanceValue(currentState.current, "#ff0"));
                    });

                    it("should have recorded three changes of state", () => {
                        expect(drawAndLog(instructionsWithPath, currentState)).toMatchSnapshot();
                    });
                });
            });
        });
    });

    describe("that describes a path that is drawn, altered and then recreated", () => {
        let recreatedPath: CurrentPath;

        beforeEach(() => {
            instructionsWithPath.moveTo(new Point(0, 0), currentState);
            instructionsWithPath.lineTo(new Point(10, 0), currentState);
            instructionsWithPath.lineTo(new Point(10, 10), currentState);
            instructionsWithPath.lineTo(new Point(0, 10), currentState);
            currentState = currentState.withCurrentState(fillStyle.changeInstanceValue(currentState.current, "#00f"));
            instructionsWithPath.lineTo(new Point(5, 5), currentState);
            recreatedPath = instructionsWithPath.recreatePath();
        });

        it("should have the same area", () => {
            expect(recreatedPath.getInstructionsToClip().area).toEqual(instructionsWithPath.getInstructionsToClip().area);
        });

        describe("and then the recreated path changes state", () => {

            beforeEach(() => {
                recreatedPath.setInitialState(defaultState.withCurrentState(fillStyle.changeInstanceValue(defaultState.current, "#00f")));
            });

            it("should not affect the state of the original", () => {
                expect(drawAndLog(instructionsWithPath, currentState)).toMatchSnapshot();
            });
        });
    });
});

describe("a set of instructions that describe a rectangle path that is drawn", () => {
    let instructionsWithPath: InstructionsWithPath;
    let currentState: InfiniteCanvasState;
    let rectangle: CanvasRectangle;

    beforeEach(() => {
        rectangle = new CanvasRectangleImpl(new MockCanvasMeasurementProvider(200, 200), {})
        currentState = defaultState;
        instructionsWithPath = InstructionsWithPath.create(currentState, rectangle);
        instructionsWithPath.rect(0, 0, 1, 1, currentState);
        instructionsWithPath.drawPath((context: CanvasRenderingContext2D) => {
            context.fill();
        }, currentState, {lineWidth: 0, lineDashPeriod: 0, shadowOffsets: []})
    });

    describe("and that then changes state", () => {

        beforeEach(() => {
            currentState = currentState.withCurrentState(fillStyle.changeInstanceValue(currentState.current, "#f00"));
        });

        it("should have recorded everything in the right order", () => {
            expect(drawAndLog(instructionsWithPath, currentState)).toMatchSnapshot();
        });
    });
});
