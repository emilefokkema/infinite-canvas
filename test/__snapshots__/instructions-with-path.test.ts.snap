// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`a set of instructions that describe a rectangle path that is drawn and that then changes state should have recorded everything in the right order 1`] = `
[
  "context.beginPath()",
  "context.moveTo(0,0)",
  "context.lineTo(1,0)",
  "context.lineTo(1,1)",
  "context.lineTo(0,1)",
  "context.lineTo(0,0)",
  "context.closePath()",
  "context.moveTo(0,0)",
  "context.fill()",
  "context.fillStyle = "#f00"",
  "context.fill()",
]
`;

exports[`a set of instructions that is also about a path that describes a path that is drawn, altered and then recreated and then the recreated path changes state should not affect the state of the original 1`] = `
[
  "context.beginPath()",
  "context.moveTo(0,0)",
  "context.lineTo(10,0)",
  "context.lineTo(10,10)",
  "context.lineTo(0,10)",
  "context.fillStyle = "#00f"",
  "context.lineTo(5,5)",
  "context.fill()",
]
`;

exports[`a set of instructions that is also about a path that receives a change of state and then receives an instruction that modifies the path and then receives a change of state on the same property as the previous change of state and then draws the path and receives another change of state on the same property should have recorded three changes of state 1`] = `
[
  "context.beginPath()",
  "context.fillStyle = "#f00"",
  "context.moveTo(0,0)",
  "context.fillStyle = "#00f"",
  "context.fill()",
  "context.fillStyle = "#ff0"",
  "context.fill()",
]
`;

exports[`a set of instructions that is also about a path that receives a change of state and then receives an instruction that modifies the path and then receives a change of state on the same property as the previous change of state should not have replaced the previous change of state 1`] = `
[
  "context.beginPath()",
  "context.fillStyle = "#f00"",
  "context.moveTo(0,0)",
  "context.fillStyle = "#00f"",
  "context.fill()",
]
`;

exports[`a set of instructions that is also about a path that receives a change of state and then receives an instruction that modifies the path should contain an instruction to begin a path, change the state and modify the path 1`] = `
[
  "context.beginPath()",
  "context.fillStyle = "#f00"",
  "context.moveTo(0,0)",
  "context.fill()",
]
`;
